prevent-broker-scale-down-if-contains-partition .md

```diff
@@ -0,0 +1,63 @@
+
+# Prevent scale down of the brokers if they contain partitions
+
+This proposal is about adding the mechanism for preventing the scale down of the brokers in the cluster if a partition is assigned to it.
+In the future, when we introduce the automatic scale up/down of the Kafka Cluster, this mechanism would help to prevent any data loss when we are scaling down the brokers when they have some partitions assigned to them.
+
+## Current situation
+
+Currently, when the number of Kafka replicas is decreased in the Kafka CR the StatefulSet will shut down the highest numbered brokers and there is no check to see if these brokers contain any topic-partition replicas.
+The [documentation](https://strimzi.io/docs/operators/latest/configuring.html#scaling-clusters-str) recommends that all topic-partition replicas are reassigned before doing this scaling action, which means that if someone tries to scale down without re-assigning the brokers, it can cause data loss.
+
+## Motivation
+
+We should definitely introduce some sort of logic which can detect if the broker which is going to be removed still contains the partitions or not.
+If the partitions are still available on the broker which is going to be scaled, then we should get some warning in the logs which will prompt users to do the reassignment and hopefully reduce the chance of data loss.
```

Comment:
People not having Cruise Control enabled will have only manual reassignment as the way to go.
I would mention that for people having Cruise Control, they could run a rebalancing in `remove-brokers` mode.
Refer to the corresponding proposal, and maybe documentation if it will be merged before this proposal got accepted.

Reply:
> we should get some warning in the logs

You cannot assume that anyone is reading the logs, so a warning achieves very little. I think we should fail the reconciliation in this case.

Reply:
@tombentley Hi, I thought we don't have to fail a reconciliation if somehow there were replicas present on broker. As per the discussion with Jakub, I thought we had to let the reconciliation take place without any scale down of brokers but the reconciliation should still happen

Reply:
Failing the reconciliation would mean anyway updating the Kafka CR with a warning condition about the error, so the user knows why it is not scaling.

Reply:
You cannot fail the reconciliation. The cluster might be in a bad state, but because the scaledown is fairly early in the reconciliation, it would cause the operator to never fix it and in turn the scale-down will never succeed. So it needs to pass:
* When there is any problem with connecting to the broker(s)
* When the partitions don't allow scale-down (because other brokers might have issues and not allow the partitions to move)